lightning/src/ln/channelmanager.rs

```diff
@@ -333,29 +337,173 @@ mod inbound_payment {
 		}
 		return Ok(PaymentPreimage(decoded_payment_preimage))
 	}
+}
+
+/// LDK has multiple reasons to generate fake short channel ids:
+/// 1) zero-conf channels that don't have a confirmed channel id yet
+/// 2) phantom node payments, to get an scid for the phantom node's phantom channel
+mod fake_scid {
+	use bitcoin::hash_types::BlockHash;
+	use bitcoin::hashes::hex::FromHex;
+	use chain::keysinterface::{Sign, KeysInterface};
+	use util::scid_utils;
+
+	use core::convert::TryInto;
+	use core::ops::Deref;
+
+	const TEST_SEGWIT_ACTIVATION_HEIGHT: u32 = 0;
+	const MAINNET_SEGWIT_ACTIVATION_HEIGHT: u32 = 481_824;
+	const MAX_TX_INDEX: u32 = 2_500;
+	const MAX_VOUT: u16 = 3;
+	const MAX_NAMESPACES: u8 = 8;
+	const NAMESPACE_ID_BITMASK: u8 = 0b111;
+
+	/// Fake scids are divided into namespaces, with each namespace having its own randomly-selected
+	/// identifier between [0..7], such that `scid_tx_index % MAX_NAMESPACES == namespace_identifier`.
+	/// This allows us to identify what namespace a fake scid corresponds to upon HTLC receipt, and
+	/// handle the HTLC accordingly.
+	pub(super) enum Namespace {
+		Phantom(u8),
+		// Coming soon: a variant for the zero-conf scid namespace
+	}
+
+	impl Namespace {
+		/// We generate "realistic-looking" random scids here, meaning the scid's block height is
+		/// between segwit activation and the current best known height, and the tx index and output
+		/// index are also selected from a "reasonable" range. We add this logic because it makes it
+		/// non-obvious at a glance that the scid is fake, e.g. if it appears in invoice route hints.
```

Comment:
I think fake_scids as implemented might fool human analysis but unlikely any automated one where the scids are matched against the UTXOs. LN channels must be p2wsh, so the binding holds only when you have intersection between the set of p2wsh and the scid. Given the low number of p2wsh among all the outputs types (1.4%), not sure the binding is really efficient..And you might accidentally intersect with public channel outputs, a clear heuristic that's the routing hint is a fake one.

I think it's good enough  as a starter though maybe in the future we can make this feature better with historical tables of the last X years of p2wsh/p2tr, sanitize out the public channels and randomly picks up a scid.

Reply:
Yea, its pretty weaksauce, but compiling in some tables of p2wsh outputs sucks too. After lightning ships taproot we should probably just increase the min height constant and call it a day.

Reply:
That's clearly an advantage that taproot mask key/script usage though time it's get widely used across the ecosystem to overlap with the fuzzy set of fake_scids..and by then routing on LN might have change a lot (trampoline, rendez-vous, etc).